File: content/examples/other/leetcode/problems/longest-increasing-subsequence/example-2/solution.js
function lengthOfLIS(nums) {
  const tails = [];
  for (const x of nums) {
    let left = 0;
    let right = tails.length;
    while (left < right) {
      const mid = left + Math.floor((right - left) / 2);
      if (tails[mid] < x) left = mid + 1;
      else right = mid;
    }
    tails[left] = x;
  }
  return tails.length;
}

if (require.main === module) {
  console.log(lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18]));
}

module.exports = { lengthOfLIS };

File: content/examples/other/leetcode/problems/longest-increasing-subsequence/example-2/solution.py
def length_of_lis(nums):
    tails = []
    for x in nums:
        left, right = 0, len(tails)
        while left < right:
            mid = left + (right - left) // 2
            if tails[mid] < x:
                left = mid + 1
            else:
                right = mid
        if left == len(tails):
            tails.append(x)
        else:
            tails[left] = x
    return len(tails)


if __name__ == "__main__":
    print(length_of_lis([10, 9, 2, 5, 3, 7, 101, 18]))
